### ISSUE TITLE
Fix calico-kube-controllers concurrent map writes

### ISSUE THREAD

**Issue by brahmini7632** [projectcalico/calico#9020]
## Expected Behavior
Calico-kube-controller pod shouldn't be crashing frequently

## Current Behavior
Calico-kube-controller crashing frequently with the fatal error of concurrent map writes

## Steps to Reproduce (for bugs)
Happening in all the calico clusters where we have large workloads running.

## Context
```
projectcalico.org/orchestrator":"k8s", "projectcalico.org/serviceaccount":"default"}, ServiceAccount:"default"} to Calico datastore
2024-06-28 03:33:48.851 [WARNING][1] cache.go 287: Value for key has changed, queueing update to reprogram key="kuberhealthy/dns-status-internal-1719545621" type=converter.WorkloadEndpointData
2024-06-28 03:33:48.851 [WARNING][1] pod_controller.go 299: No corresponding WorkloadEndpoint in cache, re-loading cache from datastore wep="kuberhealthy/dns-status-internal-1719545621"
2024-06-28 03:33:48.881 [INFO][1] pod_controller.go 324: Writing endpoint kuberhealthy/dns-status-internal-1719545621 with updated data converter.WorkloadEndpointData{PodName:"dns-status-internal-1719545621", Namespace:"kuberhealthy", Labels:map[string]string{"app":"kuberhealthy-check", "kuberhealthy-check-name":"dns-status-internal", "kuberhealthy-run-id":"xxxx", "projectcalico.org/namespace":"kuberhealthy", "projectcalico.org/orchestrator":"k8s", "projectcalico.org/serviceaccount":"default"}, ServiceAccount:"default"} to Calico datastore
fatal error: concurrent map writes
goroutine 150 [running]:
github.com/projectcalico/calico/libcalico-go/lib/backend/k8s/conversion.defaultWorkloadEndpointConverter.podToDefaultWorkloadEndpoint({}, 0xc002287688)
github.com/projectcalico/calico/libcalico-go/lib/backend/k8s/conversion.defaultWorkloadEndpointConverter.PodToWorkloadEndpoints(...)
        /go/src/github.com/projectcalico/calico/libcalico-go/lib/backend/k8s/conversion/workload_endpoint_default.go:61
github.com/projectcalico/calico/kube-controllers/pkg/converter.(*podConverter).Convert(0xc006aeb080?, {0x1fba860?, 0xc002287688?})
        /go/src/github.com/projectcalico/calico/kube-controllers/pkg/converter/pod_converter.go:95 +0x16e
github.com/projectcalico/calico/kube-controllers/pkg/controllers/pod.NewPodController.func3({0x40a352?, 0xc000656900?}, {0x1fba860, 0xc002287688})
        /go/src/github.com/projectcalico/calico/kube-controllers/pkg/controllers/pod/pod_controller.go:161 +0x2f1
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
        /go/pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/controller.go:246
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        /go/pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/shared_informer.go:972 +0xea
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
        /go/pkg/mod/k8s.io/apimachinery@v0.28.9/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0027d9f70, {0x230f600, 0xc000a2c060}, 0x1, 0xc000a38000)
```
## Your Environment
* Calico version: 3.26.2 & 3.28.0
* Orchestrator version: kubernetes
* Operating System and version: Flatcar 3510.2.5

**Comment by fasaxc** [comment]
Please provide a full log if you can, the often it will show the second stack trace for the conflicting write, which makes it a lot easier to unpick.

**Comment by fasaxc** [comment]
Think the problem might be that our `podToDefaultWorkloadEndpoint` method expects to have exclusive ownership of the `Pod` object, so, it doesn't bother to copy the labels map/annotations map.  Instead it re-uses them from the Pod and modifies them in-place.

When called from the syncer we do have exclusive ownership but the shared informer has the object in its cache so we shouldn't modify those maps.

**Comment by fasaxc** [comment]
Closing as fixed in v3.28.1 (releasing soon)/ v3.29.0.  Please open a new issue if you still see it after upgrading to one of those releases.  (I know I fixed a race, but I can't be 100% sure it was the one you hit!)

If you want to try a pre-release build of v3.28.1, there's one here: https://2024-07-28-v3-28-simmering.docs.eng.tigera.net/
